### DISCUSSION TITLE
This Serverless Function has crashed

### DISCUSSION THREAD

**Discussion by DBonbon** [vercel/next.js#47615] [Help]
### Summary

Problem:

I struggle with deploying an app that functions on my local machine. I assume my vercel configs are are wrong. But, I don't know where. 

Any help, tip would be highly appreciated. 

After I deploy the my git repo the following error dis displayed:

I'm newbie to Vercel. Kindly advise if any additional info is needed. 
```
This Serverless Function has crashed.
Your connection is working correctly.
Vercel is working correctly.
500: INTERNAL_SERVER_ERROR Code: FUNCTION_INVOCATION_FAILED ID: fra1::5s85j-1680018744610-e611cac6920c
I get the following error when deploying:

```

The log displays:
```
info  - Loaded env from /var/task/.env
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:14062:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async getRequest (/var/task/.next/server/chunks/393.js:111:17)
    at async getPage (/var/task/.next/server/chunks/393.js:70:12)
    at async getServerSideProps (/var/task/.next/server/chunks/353.js:52:102) {
  cause: Error: connect ECONNREFUSED 127.0.0.1:8081
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1487:16) {
    errno: -111,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '127.0.0.1',
    port: 8081
  },
  page: '/'
}
RequestId: 2f6f287f-ac82-441c-b1e3-c4b03ffeb6d3 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

Background:
This NextJs app fetch data from a django project. 

On my local station the app fetches data as expected with the backend repo. 

I deployed it as is without modifying the configs on my repo. 

### Additional information

```js
my local backend api is:

WAGTAIL_API_URL=http://localhost:8081/wt/api/nextjs
NEXT_PUBLIC_WAGTAIL_API_URL=/wt/api/nextjs

package.json:
{
    "name": "frontend_nextjs",
    "version": "0.1.0",
    "private": true,
    "scripts": {
        "dev": "next dev -p 3000",
        "build": "next build",
        "build:local": "IGNORE_SENTRY=1 next build",
        "start": "next start -p 3000",
        "export": "next export",
        "storybook": "start-storybook -p 3001",
        "build-storybook": "npx build-storybook",
        "fixcode": "npx prettier \"**/*(*.css|*.js|*.json|*.md|*.scss)\" --write",
        "new": "node cli/bin.js new",
        "new:container": "node cli/bin.js new --container",
        "test": "jest --watch",
        "test:ci": "npx jest --ci --runInBand --reporters=default --reporters=jest-junit"
    },
    "dependencies": {
        "@sentry/nextjs": "^7.42.0",
        "@types/react": "^18.0.28",
        "classnames": "^2.3.2",
        "i18next": "^22.4.11",
        "next": "13.2.4",
        "next-i18next": "^13.2.2",
        "prop-types": "^15.8.1",
        "react": "18.2.0",
        "react-dom": "18.2.0",
        "react-i18next": "^12.2.0"
    },
    "browserslist": {
        "production": [
            ">0.2%",
            "not dead",
            "not op_mini all"
        ],
        "development": [
            "last 1 chrome version",
            "last 1 firefox version",
            "last 1 safari version"
        ]
    },
    "husky": {
        "hooks": {
            "pre-commit": "pretty-quick --staged --pattern 'frontend/**' && jest --onlyChanged",
            "pre-push": ""
        }
    },
    "devDependencies": {
        "@storybook/addon-a11y": "^6.5.16",
        "@storybook/addon-actions": "^6.5.16",
        "@storybook/addon-backgrounds": "^6.5.16",
        "@storybook/addon-viewport": "^6.5.16",
        "@storybook/addons": "^6.5.16",
        "@storybook/builder-webpack5": "^6.5.16",
        "@storybook/manager-webpack5": "^6.5.16",
        "@storybook/react": "^6.5.16",
        "@svgr/webpack": "^6.5.1",
        "@swc/core": "^1.3.39",
        "@swc/jest": "^0.2.24",
        "@testing-library/jest-dom": "^5.16.5",
        "@testing-library/react": "^14.0.0",
        "@testing-library/user-event": "^14.4.3",
        "cross-env": "^7.0.3",
        "css-loader": "^6.7.3",
        "eslint": "8.36.0",
        "eslint-config-next": "13.2.4",
        "eslint-config-prettier": "^8.7.0",
        "husky": "^8.0.3",
        "jest": "^29.5.0",
        "jest-environment-jsdom": "^29.5.0",
        "jest-junit": "^15.0.0",
        "prettier": "^2.8.4",
        "pretty-quick": "^3.1.3",
        "react-test-renderer": "^18.2.0",
        "storybook-addon-next": "^1.7.3",
        "style-loader": "^3.3.1"
    }
}
```

**Comment by icyJoseph** [comment]
Just to make sure, the Django server is running only in your local machine?

This line:

```console
WAGTAIL_API_URL=http://localhost:8081/wt/api/nextjs
```

Makes me think that your Vercel function is trying to reach localhost:8081, within the Vercel runtime. You'd need to provide WAGTAIL_API_URL that points to your machine's server, or another server deploy of WAGTAIL.

This is further reinforced by:

```console
  cause: Error: connect ECONNREFUSED 127.0.0.1:8081
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1487:16) {
    errno: -111,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '127.0.0.1',
    port: 8081
  },
```

Which is saying that there's no target available, on the Vercel runtime, running on port 8081.

**Reply by DBonbon** [reply]
Thanks for getting back. 

I removed wt/api/nextjs and tried to deploy with:
WAGTAIL_API_URL = 127.0.0.1:8081/ localhost:8081 with or without httm;//

I then redeploy after any change. 

The error message remain the same, as specified in the logs.

**Reply by icyJoseph** [reply]
But why would it be any different.

The Django server is not running on the Vercel deployment.

From the info you're giving, it looks like you want the Vercel deployment to find you computers local server? I'm not gonna assume that. 

Are you somehow trying to start the Django server on the deployment?

As far as I remember python is not available in Vercel Node environments

**Reply by icyJoseph** [reply]
If you want the Vercel deployment to find your computer, you should look into something like ngrok, https://ngrok.com/ - working directly with your IP feels wrong.

**Reply by DBonbon** [reply]
Thanks for pointing it out. 

My goal is to link backend and frontend both from vercel. Using in a local backend is only meant to isolate the process so I can identify potential bugs. 

That is, initially I tried to mount both back-end and front-end to vercel. As it didn't work, I opted to isolate the process. 

I looked for a repo or a reference to mount wagtail/django with fronted on vercel. I didn't find. 

I'd be more than happy with any advice, reference that will allow to mount both repos directly on vercel.

**Reply by icyJoseph** [reply]
Alright, well Vercel does support python developments, furthermore there's an example you could use a first step:

- https://vercel.com/templates/python/django-hello-world
- https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/python

Once you have deployed your Django app, you'll have a deployment URL `your-project.username.vercel.app` something like that.

I have never used Wagtail, but I guess if the environment supports python, then both Django and Wagtail should just work?

I am not sure how your project is structured, but if its all under the same repository you might have some issues I guess? I cannot confirm that.

**Reply by DBonbon** [reply]
thanks for details and references. 

I had some issues with deploying my django application. I managed to deploy the above link. But had problems to move on from there. There is the issue of the DB and mainly running commands on Vercel. 

With the DB, I found some links on how to use Postgres with NextJS. But with plateforms which I didn't work before. While this seems rather a small bump, it is less clear to me how, and if, I can control the DB from Vercel. Like I can do with Heroku, D.ocean, etc'. 

I'd be curious to learn if this is possible/recommended, or if I better follow use it as front-end only server.

**Reply by icyJoseph** [reply]
Yeah I don't know much about running Django on Vercel, and what limitations there might be. If it is a serverless deployment, then I guess the project is shutdown after a request arrives?
